### ISSUE TITLE
GDScript: Add possibility to limit instantiable nodes

### ISSUE THREAD

**Issue by Apilonius64** [godotengine/godot-proposals#9799]
### Describe the project you are working on

A plugin to export customizable NSIS-Installers
- The installer configuration is hold by a special scene (-> here called installer-scene)
- There "custom nodes" (`@tool` + `extend Node` + `class_name`)

For example, the installer-scene tree may be something like this:
```
GDInstaller
|- WelcomePage
|- LicensePage
|- DeestPage
|- (...)
```

### Describe the problem or limitation you are having in your project

There is no reason why it should be possible to instance normal Nodes  as childs of Installer Nodes. In fact, i think it may be a bit confusing if the user would always be required to search a few special nodes in the "Create Node Dialog".

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a possibility to limit instantiable nodes for specific parent nodes.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

1.  via an annotation
```
@tool
extends Node

class_name MyAmazingNode
@tree_blacklist("Node1", ..) or
@tree_whitelist("Node1", ..)
```
2. Via a function

### If this enhancement will not be used often, can it be worked around with a few lines of script?

No, because there's no possibility to get the "Create Node Dialog".

### Is there a reason why this should be core and not an add-on in the asset library?

It's not possible to implement that feature as an addon.

**Comment by AThousandShips** [comment]
Do the installer nodes really need to be added in the editor and scene tree? Please explain a bit more why these nodes *need* to be added this way (instead of automatically via a plugin) and why the child nodes need to be added manually, and not automatically or based on a configuration, or need to be nodes at all

**Comment by dalexeev** [comment]
You can use [`Node._get_configuration_warnings()`](https://docs.godotengine.org/en/stable/classes/class_node.html#class-node-private-method-get-configuration-warnings) method.

![](https://github.com/godotengine/godot-proposals/assets/47700418/bbde5bfe-e0cd-4e7b-9f11-1c7409962766)

**Comment by AThousandShips** [comment]
Would this just affect the editor? I.e. would it just affect what nodes you can pick? Or would it affect code as well? What would happen if you tried to drag a node under the installer? Or paste it? Would it affect just the installer or any ancestor?

**Comment by dalexeev** [comment]
> No, because there's no possibility to get the "Create Node Dialog".

There are [`EditorInterface.popup_node_selector()`](https://docs.godotengine.org/en/latest/classes/class_editorinterface.html#class-editorinterface-method-popup-node-selector) and [`EditorInterface.popup_property_selector()`](https://docs.godotengine.org/en/latest/classes/class_editorinterface.html#class-editorinterface-method-popup-property-selector) methods. We could expose a similar method for Create Dialog, would that solve your issue?

> It's not possible to implement that feature as an addon.

I don't know what exactly you would like to do and therefore I'm not sure if it's possible to implement. But I would not be so confident in saying that this is impossible to implement. See [dalexeev/godot-inspector-code-highlighter](https://github.com/dalexeev/godot-inspector-code-highlighter) for an example. Yes, you can't currently add a custom annotation, but there are [`@export_custom`](https://docs.godotengine.org/en/latest/classes/class_%40gdscript.html#class-gdscript-annotation-export-custom), [`PROPERTY_HINT_TYPE_STRING`](https://docs.godotengine.org/en/latest/classes/class_%40globalscope.html#class-globalscope-constant-property-hint-type-string), [`_validate_property()`](https://docs.godotengine.org/en/latest/classes/class_object.html#class-object-private-method-validate-property) and [inspector plugins](https://docs.godotengine.org/en/latest/tutorials/plugins/editor/inspector_plugins.html). Maybe some of this will work?

**Comment by Apilonius64** [comment]
> Would this just affect the editor? I.e. would it just affect what nodes you can pick? Or would it affect code as well? What would happen if you tried to drag a node under the installer? Or paste it? Would it affect just the installer or any ancestor?

Yeah, i think these are the core issues of such a feature because they also may be plugin specific and may destroy some core functionalities (or would make them unnecessarily complicated)

So i think it would be better if these detail questions would be handled by the plugin and the new feature(s) would just provide the functionality to make these enhancements (implemented in a plugin) possible. These could be done, with e.g.:
- a function to register custom handlers for the "Create Node Dialog"-Open-Signal
- a general possibility to modify the scene dock
  - Maybe partially similar to #9798, the SceneDock class may provide a function to get the "Add-Node"-button, etc.

> There are [EditorInterface.popup_node_selector()](https://docs.godotengine.org/en/latest/classes/class_editorinterface.html#class-editorinterface-method-popup-node-selector) and [EditorInterface.popup_property_selector()](https://docs.godotengine.org/en/latest/classes/class_editorinterface.html#class-editorinterface-method-popup-property-selector) methods. We could expose a similar method for Create Dialog, would that solve your issue?

It would make it a lot easier to write a custom "Create-Node"-handler function if the above suggestions got implemented.

I'll update my Issue.
___
> Do the installer nodes really need to be added in the editor and scene tree? Please explain a bit more why these nodes _need_ to be added this way (instead of automatically via a plugin) and why the child nodes need to be added manually, and not automatically or based on a configuration, or need to be nodes at all

I also thought about that (it was my first approach), but then realized I started to build an own tree-style system in the inspector. I definitely agree, with my current plugin's features, it would be way simpler to just use the plugin in the inspector (eg. with a single custom Resource), so I think I'm just testing the node/tree system as a better base for future and more complex features. It may be the wrong approach.

> I don't know what exactly you would like to do and therefore I'm not sure if it's possible to implement. But I would not be so confident in saying that this is impossible to implement. See [dalexeev/godot-inspector-code-highlighter](https://github.com/dalexeev/godot-inspector-code-highlighter) for an example. Yes, you can't currently add a custom annotation, but there are [@export_custom](https://docs.godotengine.org/en/latest/classes/class_%40gdscript.html#class-gdscript-annotation-export-custom), [PROPERTY_HINT_TYPE_STRING](https://docs.godotengine.org/en/latest/classes/class_%40globalscope.html#class-globalscope-constant-property-hint-type-string), [_validate_property()](https://docs.godotengine.org/en/latest/classes/class_object.html#class-object-private-method-validate-property) and [inspector plugins](https://docs.godotengine.org/en/latest/tutorials/plugins/editor/inspector_plugins.html). Maybe some of this will work?

Not exactly what I mean, I'll come back to that when I decide to continue developing with an inspector-based solution.

**Comment by Apilonius64** [comment]
Ok, should I close the issue?

**Comment by AThousandShips** [comment]
If you're unable to achieve it with the proposed workarounds then it might be interesting to see more support, but this doesn't seem to be the correct way of solving your issue, and isn't impossible or even hard to work around in a reasonable way IMO
